ruby/lib/gherkin/query.rb

```diff
@@ -20,34 +27,42 @@ def update_feature(feature)
       store_nodes_location(feature.tags)
 
       feature.children.each do |child|
-        update_rule(child.rule) if child.rule
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_rule(feature, child.rule) if child.rule
+        update_background(feature, child.background) if child.background
+        update_scenario(feature, child.rule, child.scenario) if child.scenario
       end
     end
 
-    def update_rule(rule)
+    def update_rule(feature, rule)
       return if rule.nil?
       store_nodes_location(rule.tags)
-
       rule.children.each do |child|
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_background(rule, child.background) if child.background
+        update_scenario(feature, rule, child.scenario) if child.scenario
       end
     end
 
-    def update_background(background)
+    def update_background(parent, background)
       update_steps(background.steps)
+      parent_id = parent.respond_to?(:id) ? parent.id : parent.object_id # Cucumber::Messages::Feature doesn't have an id :(
+      @background_scopes[parent_id] = background
     end
 
-    def update_scenario(scenario)
+    def update_scenario(feature, rule, scenario)
       store_node_location(scenario)
       store_nodes_location(scenario.tags)
       update_steps(scenario.steps)
       scenario.examples.each do |examples|
         store_nodes_location(examples.tags || [])
         store_nodes_location(examples.table_body || [])
       end
+
+      @scenario_parent_locations[scenario.id] = [
+        feature.location,
+        @background_scopes[feature.object_id]&.location,
```

Comment:
can the safe navigation operators here ever return nils i.e. be needed?

Reply:
Yeah, the first item is required, but the latter three are optional. In other words, a Scenario always belongs to a Feature (line 1), but may or may not belong to a Background (line 2), a Rule (line 3), and/or a Rule Background (line 4), so we want to handle the nil case for the last three lines.

Reply:
Again, as a curiosity, you can see that we're keying the rule by its `#id`, but we have to use `#object_id` for the feature. :(